apex/contrib/torchsched/inductor/scheduler.py

```diff
@@ -0,0 +1,537 @@
+            consumer_graph.add(node)
+            consuming_dependency = set().union(*[n.get_buffer_names() for n in consumer_graph])
+
+        if consumer_graph:
+            execution_schedule = [copy.copy(producer_graph), copy.copy(consumer_graph)]
+        else:
+            execution_schedule = [copy.copy(producer_graph)]
+
+        self.execution_schedule = execution_schedule
+
+    def debug_str_short(self, node: BaseSchedulerNode) -> str:
+        """Generate short string representing scheduler node's calling function or indices."""
+        if node.is_extern() and isinstance(node.node, ir.MultiOutput):
+            kernel_str = node.node.codegen_list_tuple_access(
+                basename="getitem",
+                indices=node.node.indices,
+            )
+            return f"{node.get_name()} ({kernel_str})"
+        elif node.is_extern():
+            kernel_name = node.node.get_kernel_name() or str(node.node.op_overload)
+            return f"{node.get_name()} ({kernel_name})"
+        else:
+            return node.get_name()
+
+    def get_last_event(self, events: set[CudaEventSym]) -> CudaEventSym:
+        """Identify the latest generated CUDA event among all given events."""
+        return sorted(events, reverse=True)[0]  # CudaEventSym is total-ordering.
+
+    def schedule_multi_cuda_streams(self) -> None:
+        """Assign each fused Inductor IR nodes with the CUDA Stream to be launched to."""
+        if not self.nodes:
+            # Empty graphs are sent to compiler in very rare circumstances. Just Skip scheduling.
+            return
+
+        buf_originate: dict[str, BaseSchedulerNode] = {}
+        node_users: dict[BaseSchedulerNode, set[BaseSchedulerNode]] = collections.defaultdict(set)
+        for node in self.nodes:
+            for n in node.get_buffer_names():
+                buf_originate[n] = node
+        for node in self.nodes:
+            for d in node.unmet_dependencies:
+                assert d.name in buf_originate
+                node_users[buf_originate[d.name]].add(node)
+
+        critical_path_per_depth: dict[int, set[BaseSchedulerNode]] = collections.defaultdict(set)
+        node_depth: dict[BaseSchedulerNode, int] = collections.defaultdict(lambda: -1)
+
+        def visit(node: BaseSchedulerNode, depth: int, prev: set[BaseSchedulerNode]) -> None:
+            if node_depth[node] < depth:
+                node_depth[node] = depth
+                path = prev | {node}
+                if len(critical_path_per_depth[depth]) < len(path):
+                    critical_path_per_depth[depth] = path
+                for user in node_users[node]:
+                    visit(user, depth + 1, path)
+
+        graph_entries = [n for n in self.nodes if not n.unmet_dependencies]
+        for entry in graph_entries:
+            visit(entry, depth=1, prev=set())
+
+        max_depth, longest_critical_path = sorted(critical_path_per_depth.items(), reverse=True)[0]
```

Comment:
is there an estimate from how far from optimal this strategy is compared to e.g., flop count, measured execution time, etc.?

Reply:
No, we don't have that yet. We need a strategy to achieve it. Some other studies in the project do use measured execution time, but not for multi-stream scheduling.